File: bmmpy/solver.py
# bmmpy/exact.py
"""
Exact basis optimization helpers for bmmpy.

This module exposes an exact solver for the following objective:

- among all bases of the row space of a window, find one with minimum total row weight

The current implementation is intentionally capped and is meant for small exact
runs such as 24-row windows.
"""

from __future__ import annotations

from ._bmmpy import (
    ExactBasisResult,
    ExactBasisSolver as _NativeExactBasisSolver,
    ExactBasisSolverConfig as _NativeExactBasisSolverConfig,
    RowWindow,
)


ExactBasisSolverConfig = _NativeExactBasisSolverConfig


class ExactBasisSolver:
    __slots__ = ("max_rows", "max_states", "max_storage_bytes", "_impl")

    def __init__(
        self,
        *,
        max_rows: int = 24,
        max_states: int = (1 << 24) - 1,
        max_storage_bytes: int = 128 * 1024 * 1024,
    ) -> None:
        config = _NativeExactBasisSolverConfig()
        config.max_rows = max_rows
        config.max_states = max_states
        config.max_storage_bytes = max_storage_bytes

        self.max_rows = max_rows
        self.max_states = max_states
        self.max_storage_bytes = max_storage_bytes
        self._impl = _NativeExactBasisSolver(config)

    def __repr__(self) -> str:
        return (
            "ExactBasisSolver("
            f"max_rows={self.max_rows}, "
            f"max_states={self.max_states}, "
            f"max_storage_bytes={self.max_storage_bytes})"
        )

    def name(self) -> str:
        return self._impl.name()

    def solve(self, window: RowWindow) -> ExactBasisResult:
        return self._impl.solve(window)


__all__ = [
    "ExactBasisResult",
    "ExactBasisSolver",
    "ExactBasisSolverConfig",
]

File: python/bindings/solver.cpp
#include "bindings.hpp"
#include "bmmpy/solver/exact_basis_solver.hpp"

#include <nanobind/stl/vector.h>

namespace nb = nanobind;

namespace bmmpy::bindings {

void bind_solver(nb::module_& m) {
    nb::class_<::bmmpy::ExactBasisSolverConfig>(m, "ExactBasisSolverConfig")
        .def(nb::init<>())
        .def_rw("max_rows", &::bmmpy::ExactBasisSolverConfig::max_rows)
        .def_rw("max_states", &::bmmpy::ExactBasisSolverConfig::max_states)
        .def_rw("max_storage_bytes", &::bmmpy::ExactBasisSolverConfig::max_storage_bytes);

    nb::class_<::bmmpy::ExactBasisResult>(m, "ExactBasisResult")
        .def(nb::init<>())
        .def_rw("input_rows", &::bmmpy::ExactBasisResult::input_rows)
        .def_rw("cols", &::bmmpy::ExactBasisResult::cols)
        .def_rw("rank", &::bmmpy::ExactBasisResult::rank)
        .def_rw("enumerated_states", &::bmmpy::ExactBasisResult::enumerated_states)
        .def_rw("total_weight", &::bmmpy::ExactBasisResult::total_weight)
        .def_rw("basis_masks", &::bmmpy::ExactBasisResult::basis_masks)
        .def_rw("basis_weights", &::bmmpy::ExactBasisResult::basis_weights)
        .def_rw("transform_matrix", &::bmmpy::ExactBasisResult::transform_matrix)
        .def_rw("basis_matrix", &::bmmpy::ExactBasisResult::basis_matrix);

    nb::class_<::bmmpy::ExactBasisSolver>(m, "ExactBasisSolver")
        .def(nb::init<::bmmpy::ExactBasisSolverConfig>(),
             nb::arg("config") = ::bmmpy::ExactBasisSolverConfig{})
        .def("name", &::bmmpy::ExactBasisSolver::name)
        .def("solve", &::bmmpy::ExactBasisSolver::solve, nb::arg("window"));
}

} // namespace bmmpy::bindings